examples/schedulers/static_thread_pool.hpp

```diff
@@ -88,11 +89,275 @@ namespace example {
+            std::__nothrow_callable<Fun, Shape, Args...> &&
+            // and emplacing a tuple doesn't throw
+            noexcept(std::execution::__decayed_tuple<Args...>(std::declval<Args>()...))
+            // there's no need to advertise completion with `exception_ptr`
+          >;
+
+      template <class SenderId, class ReceiverId, class Shape, class Fun, bool MayThrow>
+        struct bulk_shared_state : task_base {
+          using Sender = std::__t<SenderId>;
+          using Receiver = std::__t<ReceiverId>;
+
+          using variant_t =
+            std::execution::__value_types_of_t<
+              Sender,
+              std::execution::env_of_t<Receiver>,
+              std::__q<std::execution::__decayed_tuple>,
+              std::__q<std::execution::__variant>>;
+
+          variant_t data_;
+          static_thread_pool& pool_;
+          Receiver receiver_;
+          Shape shape_;
+          Fun fn_;
+
+          std::atomic<std::uint32_t> finished_threads_{0};
+          std::atomic<std::uint32_t> thread_with_exception_{0};
+          std::exception_ptr exception_;
+
+          static std::pair<Shape, Shape> 
+          even_share(Shape n, std::uint32_t rank, std::uint32_t size) noexcept {
+            const auto avg_per_thread = n / size;
+            const auto n_big_share = avg_per_thread + 1;
+            const auto big_shares = n % size;
+            const auto is_big_share = rank < big_shares;
+            const auto begin = is_big_share ? n_big_share * rank
+                                            : n_big_share * big_shares +
+                                                (rank - big_shares) * avg_per_thread;
+            const auto end = begin + (is_big_share ? n_big_share : avg_per_thread);
+
+            return std::make_pair(begin, end);
+          }
+
+          std::uint32_t num_agents_required() const {
+            return std::min(shape_, static_cast<Shape>(pool_.available_parallelism()));
+          }
+
+          bulk_shared_state(
+              static_thread_pool& pool,
+              Receiver receiver, Shape shape, Fun fn) 
+            : pool_(pool)
+            , receiver_{(Receiver&&)receiver}
+            , shape_{shape}
+            , fn_{fn}
+            , thread_with_exception_{num_agents_required()}
+          {
+            this->__execute = [](task_base* t, std::uint32_t tid) noexcept {
+              auto& self = *static_cast<bulk_shared_state*>(t);
+              auto total_threads = self.num_agents_required();
+
+              if constexpr (MayThrow) {
+                try {
+                  std::visit([&](auto& tupl) -> void {
+                    std::apply([&](auto&... args) -> void {
+                      auto [begin, end] = even_share(self.shape_, tid, total_threads);
+                      for (Shape i = begin; i < end; ++i) {
+                        self.fn_(i, args...);
+                      }
+                    }, tupl);
+                  }, self.data_);
```

Comment:
If you have a member function that accepts the inner lambda here as an argument, then you can reuse it 4 times. (The lambda can always take by lvalue reference, and the `set_value` ones can `std::move` the tuple elements.)

Then, 2 of the 4 invocations would be passing the same lambda. Create two named lambdas and reuse them in the 4 invocations of the member function you add. This function gets a lot smaller and easier to maintain.